### ISSUE TITLE
[Bug]: v6.4 DataBrowserRouter with React 18's hydrateRoot

### ISSUE THREAD

**Issue by manspaniel** [remix-run/react-router#8972]
### What version of React Router are you using?

6.4.0-pre-7 + React 18.1.0

### Steps to Reproduce

Issue occurs when using React 18's `hydrateRoot` on server-rendered content, using a `<DataBrowserRouter>` on the client and a `<StaticRouter>` on the server.

Have created a [possibly bad] repro via CodeSandbox [here](https://codesandbox.io/s/react-router-6-4-pre-7-hydrate-issue-tkgnsq?file=/src/index.tsx.)

Swapping to `<BrowserRouter>` removes the error, and no hydration errors occur.

### Expected Behavior

Hydration should occur without errors when using `ReactDOMClient.hydrateRoot`.

### Actual Behavior

The following error is thrown:

```
Uncaught Error: Missing getServerSnapshot, which is required for server-rendered content. Will revert to client rendering.
    at mountSyncExternalStore (react-dom.development.js:16790:13)
    at Object.useSyncExternalStore (react-dom.development.js:17733:14)
    at useSyncExternalStore (react.development.js:1676:21)
    at useRenderDataRouter (components.tsx:77:28)
    at DataBrowserRouter (index.tsx:179:10)
    at renderWithHooks (react-dom.development.js:16311:18)
    at mountIndeterminateComponent (react-dom.development.js:20080:13)
    at beginWork (react-dom.development.js:21593:16)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:4164:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:4213:16)
```

**Comment by brophdawg11** [comment]
We're still finalizing the SSR story for the new router - but for the time being you can patch this with a one-liner: https://github.com/remix-run/react-router/pull/8986/files#diff-4e0bee20b73c1884e4a8048322e60766d2323c733cb7919be92f10b4a2a02e3aR79.  Just need to add `() => router.state` as the 3rd argument to `useSyncExternalStoreShim`.  We hope to have something released next week (likely a `6.4.0-pre.9`) with the SSR stuff close to complete.

**Comment by ryanflorence** [comment]
Like Matt said, we're still working on the static routing stuff for data, so I'm closing this :)
